mlflow/transformers.py

```diff
@@ -0,0 +1,793 @@
+
+    flavor_configuration = {
+        _TASK_KEY: task,
+        _LOGGED_TYPE_KEY: _get_instance_type(model, True),
+        _INSTANCE_TYPE_KEY: _get_instance_type(model, False),
+        _MODEL_PATH_OR_NAME_KEY: _get_base_model_architecture(model),
+    }
+
+    # If the object to be saved is a Pipeline, record the model type within the pipeline
+    if isinstance(model, Pipeline):
+        flavor_configuration.update(
+            {_PIPELINE_MODEL_TYPE_KEY: _get_instance_type(model.model, False)}
+        )
+    return flavor_configuration
+
+
+def _get_or_infer_task_type(model, task: Optional[str] = None) -> str:
+    """
+    Validates that a supplied task type is supported by the ``transformers`` library if supplied,
+    else, if not supplied, infers the appropriate task type based on the model type.
+    """
+    if task:
+        _validate_transformers_task_type(task)
+    else:
+        task = _infer_transformers_task_type(model)
+    return task
+
+
+def _infer_transformers_task_type(model) -> str:
+    """
+    Performs inference of the task type, used in generating a pipeline object based on the
+    underlying model's intended use case. This utility relies on the definitions within the
+    transformers pipeline construction utility functions.
+
+    :param model: Either the model or the Pipeline object that the task will be extracted or
+                  inferred from
+    :return: The task type string
+    """
+    from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel, Pipeline
+    from transformers.pipelines import get_task
+
+    if isinstance(model, Pipeline):
+        return model.task
+    elif isinstance(model, (PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel)):
+        # transformers provides guard conditions for potentially invalid entries
+        return get_task(model.name_or_path)
+    else:
+        raise MlflowException(
+            f"The provided model type: {type(model)} is not supported. "
+            "Supported model types are: PreTrainedModel or Pipeline."
+        )
+
+
+def _validate_transformers_task_type(task: str) -> None:
+    """
+    Validates that a given ``task`` type is supported by the ``transformers`` library.
+    """
+    from transformers.pipelines import get_supported_tasks
+
+    valid_tasks = get_supported_tasks()
+
+    if task not in valid_tasks:
+        raise MlflowException(
+            f"The task provided is invalid. '{task}' is not a supported task. "
+            f"Must be one of: {valid_tasks}"
+        )
+
+
+def _get_engine_type(model):
+    """
+    Determines the underlying execution engine for the model based on the 3 currently supported
+    deep learning framework backends: ``tensorflow``, ``torch``, or ``flax``.
+    """
+    from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel
+
+    for cls in model.__class__.__mro__:
+        if issubclass(cls, TFPreTrainedModel):
```

Comment:
Can we use `isinstance(model, TFPreTrainedModel)`?

Reply:
The actual model types that are being checked aren't instances of those ABC's. Checking with `isinstance` on the mro table entry doesn't resolve, since we're validating on classes instead of objects.

Reply:
Got it, thanks for the explanation :)